proposals/0000-type-lambda.rst

```diff
@@ -0,0 +1,287 @@
+  of ``-XScopedTypeVariables`` to bring ``k`` into scope in ``f``\'s definition. This is
+  regrettable but seems an inevitable consequence of the ``{k}`` notation.
+
+.. _`26`: https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0026-explicit-specificity.rst
+.. _`#99`: https://github.com/ghc-proposals/ghc-proposals/pull/99
+  
+* (technical) The `Visible Type Applications`_ (VTA) paper defines the behavior about what to
+  do when checking against a polytype: it says to deeply skolemize. However, eager deep
+  skolemization will spell trouble for this extension, as we need the lambdas to see
+  the ``forall``\s. The end of the Section 6.1 in the `extended VTA <https://cs.brynmawr.edu/~rae/papers/2016/type-app/visible-type-app-extended.pdf>`_ paper discusses
+  why we do eager deep skolemization: essentially, the alternative would be to do
+  type generalization at inflection points between checking and inference mode,
+  right before doing the subsumption check. Type generalization is hard in GHC, though,
+  and so the paper avoided it. In order to implement this proposal, we'll have to work
+  out how to do this.
+
+Costs and Drawbacks
+-------------------
+This is another feature to specify and maintain, and that's always a burden. It will take
+some creative thought about how to do generalization properly (last point in previous section),
+but I don't actually think the code will be all that challenging there.
+
+There is a potential confusion with as-patterns.
+
+Alternatives
+------------
+If we want to bind type variables in lambda-expressions, I think this is the only way to do it.
+We don't have to, of course, but then there will still be one area in GHC/Haskell that requires
+``Proxy``, and that's unfortunate.
+
+One alternative design would be to rearrange the extensions so that users could enable
+parts of today's ``ScopedTypeVariables`` without enabling the strange binding behavior of
+``forall``. I don't feel the need for this, myself, so I do not plan on working out this
+design, but I'm happy to accept contributions toward this end from the community. One such
+worked out design is in `this comment <https://github.com/ghc-proposals/ghc-proposals/pull/155#issuecomment-406024481>`_.
+I'm still not convinced the complication is worth it.
+
+One drawback of this proposal is that it rejects ``id @a (x :: a) = x`` if there is no
+type signature on ``id``. We could imagine extending this feature to pretend that such
+a definition comes with an implicit ``id :: forall a. a -> _`` partial type signature
+and proceeding accordingly. (The partial type signature is created from a quick syntactic
+analysis of the definition.) In this case, the definition of ``id`` would be accepted.
+However, I worry that this would be fragile as the partial-type-signature extraction would
+have to be purely syntactic. For example, would ``null @a ((_ :: a) : _) = False`` be treated
```

Comment:
How about: all type-variables have to bound if any?  I think even bad looking cases like `const @a @b x = x` aren't bad. I imagine that type-checker would generate a new type-variable for `x`, but it won't be solved, so it can error with ambiguous type error.